### PR TITLE
NativeEthernet transport for Teensy 4.1

### PR DESCRIPTION
**Pull request by KohlhardtC**

### CODE REVIEW

File: src/micro_ros_arduino.h [hunk 1]
```diff
@@ -69,4 +69,112 @@ static inline void set_microros_wifi_transports(char * ssid, char * pass, char *
...
+
+/**
+  * Base Native Ethernet implementation which could be handy if TCP becomes supported by Micro-Ros in the future
+  * as this code would be general to both TCP and UDP implementations 
+  */
+static inline micro_ros_agent_locator set_microros_native_ethernet_transports(byte mac[],IPAddress fallback_client_ip, IPAddress agent_ip , uint agent_port ) { 
+
+
+
+    #ifdef MR_DEBUG_TO_SERIAL
```
Comment (pablogs9) [inline src/micro_ros_arduino.h:98]: Please remove all the debug traces to serial port of this implementation

**Review by pablogs9 — COMMENTED** [review]
You need to rebase this PR to the latest `foxy` branch

**Comment by KohlhardtC** [comment]
@pablogs9 I believe this is ready to go now

File: .github/workflows/library_generation.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
-name: Library generator 
-on: 
+name: Library generator
```
Comment (pablogs9) [inline .github/workflows/library_generation.yml:1]: You should not be modifying this file

File: built_packages [hunk 3]
```diff
@@ -1,34 +1,34 @@
 
-https://github.com/ros2/rcpputils.git 13b04da7864c3f0afe8a6834ad84c22ca389a4f8
-https://github.com/ros2/rcl.git f96196512edd48f89eefdb6d590a7ef879053971
-https://github.com/ros2/test_interface_files.git d0ee98f449c428fc2afa888b06fea48b87dc6645
-https://github.com/ros2/rosidl_defaults.git cbc7ffd38d98c5b15823fc62b0891d2264ae4472
-https://github.com/ros2/rcl_interfaces.git ee04e046132be9d9fdcdaeff8dd54a5d0d9aa022
-https://github.com/ros2/libyaml_vendor.git 4d360d783b3018dfaed6d9ab9c736b720c109702
-https://github.com/ros2/rosidl_dds.git e88b1d0e62a2dca0788142cf1fb266a3a3c3d7dc
+https://github.com/eProsima/Micro-XRCE-DDS-Client.git 8b86e86848811428c1a2bdf7acd84ed9215c74e6
```
Comment (pablogs9) [inline built_packages:2]: You should not be modifying this file

File: available_ros2_types [hunk 4]
```diff
@@ -1,234 +1,234 @@
...
-rcl_interfaces/DescribeParameters.srv
-lifecycle_msgs/Transition.msg
-lifecycle_msgs/State.msg
-lifecycle_msgs/TransitionEvent.msg
-lifecycle_msgs/TransitionDescription.msg
-lifecycle_msgs/GetState.srv
-lifecycle_msgs/GetAvailableStates.srv
-lifecycle_msgs/GetAvailableTransitions.srv
-lifecycle_msgs/ChangeState.srv
+tf2_msgs/LookupTransform.action
```
Comment (pablogs9) [inline available_ros2_types:2]: You should not be modifying this file

File: examples/micro-ros_publisher_native_ethernet/micro-ros_publisher_native_ethernet.ino [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+#include <rclc/rclc.h>
+#include <rclc/executor.h>
+
+#include <std_msgs/msg/int32.h>
+
+
+#ifndef ARDUINO_TEENSY41
+#error This example has only been tested on Teensy 4.1
+#endif 
+
```
Comment (pablogs9) [inline examples/micro-ros_publisher_native_ethernet/micro-ros_publisher_native_ethernet.ino:15]: Remove white spaces

File: src/micro_ros_arduino.h [hunk 6]
```diff
@@ -69,4 +69,81 @@ static inline void set_microros_wifi_transports(char * ssid, char * pass, char *
...
+	int port;
+}; 
+
+
+/**
+  * Base Native Ethernet implementation which could be handy if TCP becomes supported by Micro-Ros in the future
+  * as this code would be general to both TCP and UDP implementations 
+  */
+static inline micro_ros_agent_locator set_microros_native_ethernet_transports(byte mac[],IPAddress fallback_client_ip, IPAddress agent_ip , uint agent_port ) { 
+
```
Comment (pablogs9) [inline src/micro_ros_arduino.h:95]: Remove white spaces

File: src/micro_ros_arduino.h [hunk 7]
```diff
@@ -69,4 +69,81 @@ static inline void set_microros_wifi_transports(char * ssid, char * pass, char *
...
+      		while (true) {
+        		delay(1); // do nothing, no point running without Ethernet hardware
+      		}
+    	}
+    	if (Ethernet.linkStatus() == LinkOFF) {
+            // No ethernet link detected
+   		}
+
+    	// try to congifure using IP address instead of DHCP:
```
Comment (pablogs9) [inline src/micro_ros_arduino.h:113]: Indentation

File: src/micro_ros_arduino.h [hunk 8]
```diff
@@ -69,4 +69,81 @@ static inline void set_microros_wifi_transports(char * ssid, char * pass, char *
...
+   		Ethernet.begin(mac, fallback_client_ip);
+  	} else {
```
Comment (pablogs9) [inline src/micro_ros_arduino.h:115]: else without functionality

File: src/uxr/client/config.h [hunk 9]
```diff
@@ -62,4 +62,7 @@
 #define UXR_CONFIG_SHARED_MEMORY_STATIC_MEM_SIZE       10
 #endif
 
+#define UCLIENT_TWEAK_XRCE_WRITE_LIMIT
```
Comment (pablogs9) [inline src/uxr/client/config.h:65]: You should not be modifying this file

File: src/uxr/client/core/session/create_entities_bin.h [hunk 10]
```diff
@@ -32,13 +32,14 @@ extern "C"
...
  *        As a result of the reception of this submessage, the Agent will create an XRCE Participant according to
  *        the binary provided in the CREATE submessage.
- * @param session       A uxrSession structure previously initialized.
- * @param stream_id     The output stream identifier where the CREATE submessage will be buffered.
- * @param object_id     The identifier of the XRCE Participant.
- * @param domain_id     The identifier of the Domain to which the XRCE Participant belongs.
- * @param mode          The set of flags that determines the entity creation mode.
- *                      The Creation Mode Table describes the entities creation behaviour according to the
- *                      `UXR_REUSE` and `UXR_REPLACE` flags.
+ * @param session               A uxrSession structure previously initialized.
```
Comment (pablogs9) [inline src/uxr/client/core/session/create_entities_bin.h:35]: You should not be modifying this file

File: src/uxr/client/profile/multithread/multithread.h [hunk 11]
```diff
@@ -52,32 +52,16 @@ typedef struct uxrMutex
 #endif // ifdef WIN32
 } uxrMutex;
```
Comment (pablogs9) [inline src/uxr/client/profile/multithread/multithread.h:53]: You should not be modifying this file

File: src/uxr/client/profile/multithread/multithread.h [hunk 12]
```diff
@@ -98,18 +82,18 @@ UXRDLLAPI void uxr_unlock(
 #define UXR_UNLOCK_STREAM_ID(session, stream_id) uxr_unlock(uxr_get_stream_mutex_from_id(session, stream_id))
 
 #define UXR_LOCK_ALL_INPUT_STREAMS(session) \
-    for (uint8_t i = 0; i < session->streams.input_best_effort_size; \
-            ++i){ uxr_lock(&session->streams.input_best_effort[i].mutex); } \
-    for (uint8_t i = 0; i < session->streams.input_reliable_size; ++i){ uxr_lock( \
-                                                                            &session->streams.input_reliable[i].mutex); \
+    for (uint8_t i = 0; i < session->streams.input_best_effort_size; ++i){ \
```
Comment (pablogs9) [inline src/uxr/client/profile/multithread/multithread.h:85]: You should not be modifying this file

**Review by pablogs9 — CHANGES_REQUESTED** [review]
Also, this PR should not modify any of the `.a` files

**Comment by KohlhardtC** [comment]
I only modified mico_ros_arduino.h, the example, and the native Ethernet transport implementation. It is beyond my expertise to understand how all the other files were modified. Maybe due to the rebase?  Some automated process? 

Do you have a published guide regarding code style preference?

Chris

**Comment by KohlhardtC** [comment]
@pablogs9 - I wanted to try one more time to see if you are willing to assist with this pull request. The changes I made are very simple and I believe only 1 modified file and 2 or 3 additions. The additional modified files you noted were not modified by me, and I'm entirely confused how that could happen. If it's easier, I'm happy to try and tackle this another way. To be honest, this pull request process has been a lot more work than the feature implementation itself and I don't want to keep beating my head against the wall on this if there isn't mutual interest in helping me work through issues with the pull request.

**Comment by pablogs9** [comment]
I have done some modification and pushed to your branch:
1. Let me know if you are ok with them.
2. Please test them on the board and confirm that everything works

Then I will merge and backport to `galactic` and `rolling`

**Review by pablogs9 — APPROVED** [review]

**Comment by pablogs9** [comment]
@KohlhardtC friendly ping

**Comment by KohlhardtC** [comment]
For reasons I don't understand, I'm not seeing revision history on these files any more and therefore cannot comment specifically on the modifications you've made (e.g. `git log native_ethernet_transport.cpp` is only showing you as an author now). That said, nothing is jumping out at me as obviously problematic. 

Unfortunately I am no longer able to get the example working using the most recent dockerized foxy agent. The example does not seem to get past this line:
` RCCHECK(rclc_node_init_default(&node, "micro_ros_arduino_node", "", &support));`

The dockerized agent only produces output of:

```
sudo docker run -it --rm --net=host -v /dev:/dev -v /dev/shm:/dev/shm --privileged microros/micro-ros-agent:foxy udp4 --port 8888 --dev /dev/ttyACM0 -v6
[1624540719.859187] info     | UDPv4AgentLinux.cpp | init                     | running...             | port: 8888
[1624540719.859313] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 6


```

**Comment by pablogs9** [comment]
I made no modifications just copy the files you added and removed the changes that are not proper. Are you able to recheck the transport and make it work?

So sorry for the inconvenience...

**Comment by carmiac** [comment]
Any idea when this PR will be pulled in?

**Comment by pablogs9** [comment]
When it can be tried by some user with the proper hardware

**Comment by carmiac** [comment]
Ok, awesome.  I'll work on getting this to build and test it.  Do you have any tips for setting up a build?

**Comment by KohlhardtC** [comment]
@pablo - would it help if I sent you this board to test with? Since you’re doing merges it seems like it would be best if you could verify the PR works after the merge.

Chris

**Comment by pablogs9** [comment]
@KohlhardtC That would help, yes

**Comment by carmiac** [comment]
I have just confirmed that this works, both the demo and in my own application that uses custom messages, rmw_uros_ping_agent and rmw_uros_sync_session.  One thing that I canged was to add a try_dhcp flag to set_microros_native_ethernet_udp_transports and set_microros_native_ethernet_transports, and then changed the first few lines of set_microros_native_ethernet_transports to:
` int eStatus = 0;

  if (try_dhcp)
	eStatus = Ethernet.begin(mac);`

That way I didn't have to wait for 60 seconds every time I started up since I'm not using DHCP.

**Comment by carmiac** [comment]
Also, I would be happy to run Teensy 4.1/Ethernet tests in the future.

**Comment by KohlhardtC** [comment]
@pablo - If you send me an address I'll get a few of these to you

**Comment by pablogs9** [comment]
We have  merged this: https://github.com/micro-ROS/micro_ros_arduino/pull/603/files#diff-7f3024ab66ceb49addbfc8ad503b8ecf8b64d3e1aab251942c648ad9ac3729e5

Can we check if those transports are the same?

**Comment by carmiac** [comment]
They just about are. The only difference is that the Teensy version should include NativeEthernet.h instead of LwIP.h and STM32Ethernet.h

**Comment by pablogs9** [comment]
Can we close this PR and open another with some conditional compilations to select whatever is needed for teensy?

**Comment by carmiac** [comment]
Yes, that makes sense.  I'll work on creating that PR.

**Comment by pablogs9** [comment]
I'll close this when the new PR is ready

**Comment by pablogs9** [comment]
Superseed by https://github.com/micro-ROS/micro_ros_arduino/pull/625

**Pull request closed without merging**